activesupport/lib/active_support/concurrency/load_interlock_aware_monitor.rb

```diff
@@ -6,12 +6,24 @@ module ActiveSupport
   module Concurrency
     # A monitor that will permit dependency loading while blocked waiting for
     # the lock.
-    class LoadInterlockAwareMonitor < Monitor
+    class LoadInterlockAwareMonitor < Monitor # :nodoc:
+      EXCEPTION_NEVER = { Exception => :never }.freeze
+      private_constant :EXCEPTION_NEVER
+
       # Enters an exclusive section, but allows dependency loading while blocked
       def mon_enter
         mon_try_enter ||
           ActiveSupport::Dependencies.interlock.permit_concurrent_loads { super }
       end
+
+      def synchronize
+        Thread.handle_interrupt(EXCEPTION_NEVER) { mon_enter }
+        begin
+          yield
+        ensure
+          Thread.handle_interrupt(EXCEPTION_NEVER) { mon_exit }
```

Comment:
I don't think this is safe: an interrupt could occur just before calling `Thread.handle_interrupt`, unwinding this method without ever calling `mon_exit`

Reply:
Hmm, I see. So what do you think of the following fix? 

```diff 
--- a/activesupport/lib/active_support/concurrency/load_interlock_aware_monitor.rb
+++ b/activesupport/lib/active_support/concurrency/load_interlock_aware_monitor.rb
@@ -6,9 +6,10 @@ module ActiveSupport
   module Concurrency
     # A monitor that will permit dependency loading while blocked waiting for
     # the lock.
-    class LoadInterlockAwareMonitor < Monitor # :nodoc:
+    class LoadInterlockAwareMonitor < Monitor
       EXCEPTION_NEVER = { Exception => :never }.freeze
-      private_constant :EXCEPTION_NEVER
+      EXCEPTION_IMMEDIATE = { Exception => :immediate }.freeze
+      private_constant :EXCEPTION_NEVER, :EXCEPTION_IMMEDIATE
 
       # Enters an exclusive section, but allows dependency loading while blocked
       def mon_enter
@@ -17,11 +18,16 @@ def mon_enter
       end
 
       def synchronize
-        Thread.handle_interrupt(EXCEPTION_NEVER) { mon_enter }
-        begin
-          yield
-        ensure
-          Thread.handle_interrupt(EXCEPTION_NEVER) { mon_exit }
+        Thread.handle_interrupt(EXCEPTION_NEVER) do
+          mon_enter
+
+          begin
+            Thread.handle_interrupt(EXCEPTION_IMMEDIATE) do
+              yield
+            end
+          ensure
+             mon_exit
+          end 
```

Reply:
Yeah, that looks right to me.

Unfortunate it looks so ugly, and pollutes the call stack, but I think it's the only way to implement it safely. I wish there was built-in syntax to make this cleaner (and harder to get wrong).

Reply:
Thanks for your review! I fixed. 

>  I wish there was built-in syntax to make this cleaner  

Agree. We should create a proposal to bugs.ruby-lang ;)